### ISSUE TITLE
Proposal: Signing Key Rotation (on AWS)

### ISSUE THREAD

**Issue by gshively11** [open-policy-agent/opa#2757]
We have a need to rotate the signing keys used to verify OPA bundles on a periodic basis, in an automated manner.  We also run primarily on AWS.  We propose a new OPA feature to support [AWS KMS Asymmetric Keys for digital signatures](https://aws.amazon.com/blogs/security/digital-signing-asymmetric-keys-aws-kms/), whereby OPA will do the following:

- Call the KMS API to sign bundle files when building bundles. AWS Credentials will be needed during this step, similar to the s3 service creds feature.
- When running as REPL/server, call the KMS API to retrieve the public key(s) on load. Cache it and re-use it until bundle signature verification fails. Call KMS API for a new public key and re-attempt signature verification (presumably because the CMK has been rotated and a bundle with a new signature has been received). AWS Credentials will also be needed during this step to communicate with the KMS API.

Considerations/Limitations:

- Users will need to create an [asymmetric CMK](https://docs.aws.amazon.com/kms/latest/developerguide/symm-asymm-concepts.html)
- Users may choose to create a [CMK Alias](https://docs.aws.amazon.com/kms/latest/developerguide/programming-aliases.html) if they don't want to update key ids in their configuration (recommended).
- Users will need to implement a pipeline that will be triggered when the CMK is rotated, which will rebuild all the bundles and sign them with the new private key. Fairly straightforward using AWS Lambda and AWS CloudWatch Events.
  - The pipeline will update the CMK Alias only after all the bundles have been successfully rebuilt and published, thereby ensuring that the OPA servers aren't stuck with policy they can't verify (because they'll still get the original public key until the alias is rotated, and the time between new policy publish and alias change will be very short).
- Using asymmetric keys minimizes the network dependency on AWS services while the OPA server is running.  Public keys just need to be fetched once at startup, and then only when a key rotation occurs.

Proposed interface changes (please suggest better names/structure, these are a bit slapdash):

`opa build`
- Add flag `--aws-signing-key-id arn:aws:kms:us-west-2:111122223333:alias/MyOpaSigningKey`.
- Add flag `--aws-credentials [metadata|environment]`.  Just like the s3 signing credentials for services.
- Add flag `--aws-credentials-iam-role [iam_role_to_assume]`.  Just like the s3 signing credentials for services.
- Add flag `--aws-credentials-region [the_aws_region]` Just like the s3 signing credentials for services.

`opa run`
- Add config/flag `--set=keys.my_key_name.aws_key_id=[key_id/arn/alias]

We would greatly appreciate your feedback on this approach. Please let us know if you see a better direction to take this (we're still pretty new to OPA). If this seems like a valuable addition to OPA core, or as a plugin, we would love to contribute the implementation. Thanks!

**Comment by patrick-east** [comment]
I'm curious if having the ability to specify a URL for a public key would help solve part of this problem. As-is we have to have OPA configured with the key, but if we instead allow for pointing at a URL and having OPA periodically fetch it I think we'd likely cover the verification side of this use-case (and allow for other similar models as azure, gcp, and likely most other public clouds offer the same type of service).

Then for the signing part I'm a little bit hesitant to support all of those flags (and the likely bloat of 2x additional similar but not quite the same ones for the azure and gcp versions that will eventually follow this one). I wonder if there is something we could do to make it easier to sign bundles without adding in the custom code to OPA. Obviously much of the API can be re-used by a golang application that uses OPA as a library, but I'm not sure how clean that is if someone just wanted to use a custom encryption provider. There might be something there we could do to simplify that and allow customized builds.

@tsandall thoughts?

**Comment by tsandall** [comment]
> I'm curious if having the ability to specify a URL for a public key would help solve part of this problem. As-is we have to have OPA configured with the key, but if we instead allow for pointing at a URL and having OPA periodically fetch it I think we'd likely cover the verification side of this use-case (and allow for other similar models as azure, gcp, and likely most other public clouds offer the same type of service).

I'm not opposed to this but it's worth pointing out that the [Discovery](https://www.openpolicyagent.org/docs/latest/management/#discovery) feature could be used to rotate verification keys.

> Then for the signing part I'm a little bit hesitant to support all of those flags (and the likely bloat of 2x additional similar but not quite the same ones for the azure and gcp versions that will eventually follow this one). I wonder if there is something we could do to make it easier to sign bundles without adding in the custom code to OPA. Obviously much of the API can be re-used by a golang application that uses OPA as a library, but I'm not sure how clean that is if someone just wanted to use a custom encryption provider. There might be something there we could do to simplify that and allow customized builds.

I'm curious how this would be used...would the `build` and `sign` commands be running inside of a controlled environment like a CI pipeline? If so, I wonder what an integration would look like today. In other words, how much work is it to wrap `opa build` with a script that uses some AWS CLI to fetch the signing keys, run `opa build` accordingly, etc. If that's viable, I'd be inclined to try that out first and see how well it works--and then use whatever we learn to improve OPA.

**Comment by gshively11** [comment]
> I'm not opposed to this but it's worth pointing out that the Discovery feature could be used to rotate verification keys.

How would this work if we wanted to rotate the signing key used to sign the discovery bundle itself?  Thinking out loud...something like this?

1. Start OPA w/ with discovery and a configured key that's retrieved (prior to starting OPA) from wherever the latest signing key is stored.
2. OPA retrieves discovery bundle (signed with latest key), verifies it, and reconfigures OPA.
3. OPA downloads other bundles and uses whatever key the discovery config specifies to verify.
4. Discovery bundle is updated to add a new public key (with a different name?), but is still signed with the existing key.
5. OPA receives new discovery bundle and adds the new key.
6. New bundles (versioned, or at a different endpoint?) signed with new key are uploaded.
6. Once discovery bundle with new key is propagated to all OPA instances, discovery config is updated to instruct all bundles to grab the new version and to use new key for verification.
7. OPA retrieves bundles and uses the new signing key that was propagated in discovery to verify.

Can discovery change the signing key that's used to verify itself?  I thought I read something about the original discovery configuration being immutable to avoid bricking an OPA instance.  I'm confused as to how the key used to verify the discovery bundle can be rotated without restarting OPA.  But, assuming that something like the above flow works and we can rotate the discovery bundle signing key like this, then I think like this approach better than retrieving the public key as a separate step. Although it does require a greater degree of control over the discovery bundle...  With that said, actually signing the bundles is still an issue, as you'll see in the next section.

> how much work is it to wrap opa build with a script that uses some AWS CLI to fetch the signing keys

The private key generated when creating a KMS Asymmetric Key is never exportable (except as a part of the initial creation command).  To use the private key to sign something, you have to pass the content to the KMS API and it returns the signature.  This makes it really secure, but it also means that OPA would have to call the KMS API directly when running `opa build` or `opa sign`.  I don't see a clear way to wrap OPA to make this happen.

**Comment by tsandall** [comment]
> Can discovery change the signing key that's used to verify itself?

Currently you'd have to restart OPA to change the key used to verify discovery bundles. That said, the restriction on changes to Discovery configuration is something that comes up from time-to-time. We could make it optional (defaulting to restricted) for users that understand the risk.

> The private key generated when creating a KMS Asymmetric Key is never exportable (except as a part of the initial creation command). To use the private key to sign something, you have to pass the content to the KMS API and it returns the signature. This makes it really secure, but it also means that OPA would have to call the KMS API directly when running opa build or opa sign. I don't see a clear way to wrap OPA to make this happen.

That's useful, thanks! It sounds like KMS integration in OPA would be unavoidable then. I think the question is whether we build this into `opa` in the short term or experiment with a plugin (separate build) like we do for Envoy and ext_authz.

**Comment by gshively11** [comment]
I'm happy to try experimenting with a plugin first.  I still don't have a good understanding of the plugin architecture, but experimenting should help me learn :).  I'll report back in a bit on that.

> restriction on changes to Discovery configuration ... We could make it optional

I think that's a very good idea, since restarting OPA means caches are cleared and bundles have to be re-downloaded before it's usable again.  Safeguard could probably be added, such that automatic rollbacks happen if a bad discovery bundle/change is ingested.  Is there an issue tracking that feature request already?  If not I'm happy to create one.

**Comment by tsandall** [comment]
> I'm happy to try experimenting with a plugin first. I still don't have a good understanding of the plugin architecture, but experimenting should help me learn :). I'll report back in a bit on that.

The existing plugin architecture would work for bundle verification but probably not signing (since the latter requires integration in the `cmd` package and the `bundle` package where signing is implemented). Currently the `bundle` package assumes there is only one implementation of signing--there are on interfaces for signer or ways to implement custom signing. We would need to change that.

> I think that's a very good idea, since restarting OPA means caches are cleared and bundles have to be re-downloaded before it's usable again. Safeguards could probably be added, such that automatic rollbacks happen if a bad discovery bundle/change is ingested. Is there an issue tracking that feature request already? If not I'm happy to create one.

If we add this, I'd probably avoid rollback support (at least initially) because it'll be difficult/impossible to tell whether a change should be rolled back. Feel free to file an issue for at least exposing a startup configuration option that lets discovery self-modify.

**Comment by gshively11** [comment]
I'm ready to take a pass at this now. After thinking about it for a bit, I want to avoid a solution that requires an update to any bundles to rotate the signing key. Due to our architecture (many, separately maintained bundles, including discovery), requiring global updates to all discovery bundles is tricky from an operational perspective. By [using AWS KMS aliases](https://docs.aws.amazon.com/kms/latest/developerguide/kms-alias.html), we can rotate the keys behind the alias, avoiding the need to change anything in the bundles.  Now, the question is, how should I go about implementing this in OPA?

Now that I have more experience with plugins and the OPA codebase, I see a couple of options. Although no matter what, I'll need to refactor the bundling to use interfaces, so that different implementations can be provided.

1) I keep everything out of the OPA codebase (besides the interface refactoring) and just add this code in our custom golang wrapper. This has the downside of no upstream contributions for a feature that I think others on AWS would find useful.
2) I create a custom build (opa-aws-plugin), similar to [opa-envoy-plugin](https://github.com/open-policy-agent/opa-envoy-plugin). This keeps OPA core free of the new AWS-related code.  I see a couple issues with this approach though. First, I may need to duplicate and tweak the build and sign commands.  I'll need new args that the original commands don't support.  There might be a way to refactor the original commands in OPA core and reduce the duplication, but not sure. Also, there's the additional operational overhead of managing/syncing multiple versions of OPA.  Although this is already done for opa-envoy-plugin, so maybe y'all have it mostly automated already?  One big advantage of this approach is that we could probably justify including the golang AWS SDK as a dependency, which avoids the effort of replicating bits of functionality from the SDK to avoid the dependency, which seems like the path taken thus far in OPA core.
3) I modify opa directly to add the new functionality, which means additional CLI args for build/sign and additional AWS-specific code. I'm assuming you would not want the AWS SDK included as a dependency in OPA core (since it was avoided previously), which makes this option seems a bit less appealing.

What are your thoughts @tsandall and @patrick-east? I'm game to try any of these, or other options if you think of them, as long as I can avoid doing discovery bundle updates.

**Comment by gshively11** [comment]
Now supported with changes in #3336
